File: server/src/main/java/by/brawl/ws/spell/Uppercut.java
package by.brawl.ws.spell;

import by.brawl.ws.holder.BattlefieldHolder;
import by.brawl.ws.holder.HeroHolder;

import java.util.Arrays;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Set;

public class Uppercut implements SpellLogic {
	private String id = "6";
	private Boolean targetable = true;
	private Set<Integer> myTargets = new HashSet<>();
	private Set<Integer> enemyTargets = new HashSet<>(Arrays.asList(1));

	@Override
	public BattlefieldHolder cast(BattlefieldHolder battlefieldHolder,
								  String senderId,
								  Integer victimPosition,
								  Boolean forEnemy) {

		List<HeroHolder> sourceList = battlefieldHolder.getBattleHeroes(senderId, true);
		LinkedList<HeroHolder> modifiedHeroesList = new LinkedList<>(sourceList);
		HeroHolder movingHero = modifiedHeroesList.get(4);
		modifiedHeroesList.addFirst(movingHero);
		modifiedHeroesList.removeLast();
		sourceList.clear();
		sourceList.addAll(modifiedHeroesList);
		return battlefieldHolder;
	}

	@Override
	public String getId() {
		return id;
	}

	@Override
	public Boolean getTargetable() {
		return targetable;
	}

	@Override
	public Set<Integer> getMyTargets() {
		return myTargets;
	}

	@Override
	public Set<Integer> getEnemyTargets() {
		return enemyTargets;
	}
}

File: server/src/main/java/by/brawl/ws/spell/Uppercut.kt
package by.brawl.ws.spell

import by.brawl.ws.holder.BattlefieldHolder
import by.brawl.ws.holder.HeroHolder
import by.brawl.ws.huihui.impact.Impact
import java.util.*

class Uppercut {

//    override val id = "6"
//    override val targetable = true
//    override val myTargets = HashSet<Int>()
//    override val enemyTargets = hashSetOf(1)

    companion object {
        fun cast(battlefieldHolder: BattlefieldHolder,
                 senderId: String,
                 victimPosition: Int?,
                 forEnemy: Boolean?): BattlefieldHolder {

            val sourceList = battlefieldHolder.getBattleHeroes(senderId, true).toMutableList()
            val modifiedHeroesList = LinkedList(sourceList)
            val movingHero = modifiedHeroesList[victimPosition!! - 1]
            modifiedHeroesList.addFirst(movingHero)
            modifiedHeroesList.removeLast()
            sourceList.clear()
            sourceList.addAll(modifiedHeroesList)
            return battlefieldHolder
        }

        fun newCast(battlefieldHolder: BattlefieldHolder, impacts: Map<HeroHolder, List<Impact>>): BattlefieldHolder {

        }
    }
}